pkg/openstack/routes.go

```diff
@@ -54,19 +64,98 @@ type Routes struct {
-		allowedAddressPairs: allowedAddressPairs,
+		network:                 network,
+		os:                      os,
+		atomicRoutes:            atomicRoutes,
+		allowedAddressPairs:     allowedAddressPairs,
+		autoConfigSecurityGroup: autoConfigSecurityGroup,
 	}, nil
 }
 
+func (r *Routes) getOrCreateNodeSecurityGroup(clusterName string) error {
+	r.Lock()
+	defer r.Unlock()
+	sgName := fmt.Sprintf("k8s-node-sg-for-cluster-%s", clusterName)
+	sgId, err := secgroups.IDFromName(r.network, sgName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			mc := metrics.NewMetricContext("security_group", "create")
+			klog.V(2).InfoS("Create node security group", "name", sgName)
+			group, err := groups.Create(r.network, groups.CreateOpts{Name: sgName}).Extract()
+			if mc.ObserveRequest(err) != nil {
+				return err
+			}
+			sgId = group.ID
+		} else {
+			return err
+		}
+	}
+	r.nodeSecurityGroupId = sgId
+	return nil
+}
+
+// checkSecurityGroupRulesExist check whether the ingress security group rules that related to the podCidr and the node (nodeName) are existing,
+// these security group rules use to ensure other nodes permit the traffic from this node (nodeName) or this node's pods pass through.
+func (r *Routes) checkSecurityGroupRulesExist(podCidr string, nodeName types.NodeName, existingRules []rules.SecGroupRule) (bool, bool, error) {
+	nodes, err := r.os.nodeInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return false, false, err
+	}
+	sgRuleForPodCidrFound := false
+	sgRuleForNodeAddressFound := false
+	podPrefix, _ := netip.ParsePrefix(podCidr)
+	nodeAddr := getAddrByNodeName(nodeName, podPrefix.Addr().Is6(), nodes)
+	nodeIp, _ := netip.ParseAddr(nodeAddr)
+
+	for _, rule := range existingRules {
+		sgPrefix, _ := netip.ParsePrefix(rule.RemoteIPPrefix)
+		if sgPrefix.Overlaps(podPrefix) && podPrefix.Bits() >= sgPrefix.Bits() {
+			sgRuleForPodCidrFound = true
+		}
+		if rule.RemoteGroupID == r.nodeSecurityGroupId || sgPrefix.Contains(nodeIp) {
+			sgRuleForNodeAddressFound = true
+		}
+		if sgRuleForPodCidrFound && sgRuleForNodeAddressFound {
+			break
+		}
+	}
+	return sgRuleForPodCidrFound, sgRuleForNodeAddressFound, nil
+}
+
+// checkPortSecurityRules check the port's security rules (SecurityGroups and AllowAddressPairs) whether are valid
+func (r *Routes) checkPortSecurityRules(port *PortWithPortSecurity, route routers.Route) bool {
+	if r.autoConfigSecurityGroup {
+		// check if the node security group bind to the port
+		nodePortBindSecurityGroup := false
+		for _, sg := range port.SecurityGroups {
+			if sg == r.nodeSecurityGroupId {
+				nodePortBindSecurityGroup = true
+				break
+			}
+		}
+		if !nodePortBindSecurityGroup {
+			return false
+		}
+	}
+
+	if r.allowedAddressPairs {
+		// check whether the related AllowAddressPair is existing
+		for _, addrPair := range port.AllowedAddressPairs {
+			if addrPair.IPAddress == route.DestinationCIDR {
+				return true
+			}
+		}
+		return false
+	}
+	return true
```

Comment:
```suggestion
	// check whether the related AllowAddressPair is existing
	if r.allowedAddressPairs && !slices.Contains(port.AllowedAddressPairs, route.DestinationCIDR) {
		return false
	}

	return true
```

Reply:
`port.AllowedAddressPairs` is a struct object while `route.DestinationCIDR` is a string object. So, `slices.Contains` cannot directly be applied.